Add deployment instructions for kubernetes deploy (#162)
opea-project/GenAIEval: evals/benchmark/README.md

## Table of Contents

> - [Workload Deployment in Kubernetes](#workload-deployment-in-kubernetes)

Reviewer: I was under the impression we keep "deployment" under GenAIExamples, because it is essential to have a single docker node or a kubernetes cluster of one or more nodes to run your GenAIExamples, or to benchmark. ..so move deployment to GenAIExamples.  Which also means you need to cover single docker node - bare metal or VM and give them link to learn how to download docker
Something along the below lines

sudo apt update -y
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
DEBIAN_FRONTEND=noninteractive sudo -E add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" -y
sudo apt update -y
DEBIAN_FRONTEND=noninteractive sudo -E apt install docker-ce docker-ce-cli containerd.io -y
sudo usermod -aG docker $USER

Author: Yes, once we achieve alignment on the out-of-the-box (OOB) flows for benchmarking in Kubernetes, we can proceed with reorganizing other sections. The 1/2/4/8 manifests pertain specifically to the current Kubernetes deployment.

  - [Prerequisites](#prerequisites)
  - [Deployment](#deployment)
    - [1. Deploy with CPUs only](#1-deploy-with-cpus-only)
    - [2. Deploy with Gaudi Devices](#2-deploy-with-gaudi-devices)
    - [3. Deploy with Multiple Instances](#3-deploy-with-multiple-instances)
    - [4. Deploy with a Specific Namespace](#4-deploy-with-a-specific-namespace)
    - [5. Deploy with Proxy Support](#3-deploy-with-proxy-support)
  - [Destroy](#destroy)
- [Installation](#installation)
- [Usage](#usage)
- [Configuration](#configuration)
  - [Test Suite Configuration](#test-suite-configuration)
  - [Test Cases](#test-cases)

> ## Workload Deployment in Kubernetes

Reviewer: A little confusing, you use plural, as in CPU**s**, Gaudi instance**s**, why call out separately "multiple instances"?  Let the Kubespray and kubeadm instructions help the user with multi-node cluster setup. Let us also stick to nodes for the compute resources, and mention they could be bare metal machines with OS or virtual machines. Let us reserve instances for container and/or pod instances.

Your multi-node cluster section is to educate the user about model download overhead to multiple nodes can be reduced and secondly how to control where some workloads can land with labels.

Author: s/Deploy with Multiple Instances/Configure Multiple Pod Instances.

"Your multi-node cluster section is to educate the user about model download overhead to multiple nodes can be reduced and secondly how to control where some workloads can land with labels." For `replicaCount`,  do you suggest removing it? While enabling Horizontal Pod Autoscaler (HPA) for scaling is important, it falls under tuning rather than being out-of-the-box (OOB) functionality. I recommend retaining the replicaCount field as a reference for others, especially when we remove the 1/2/4/8 manifests."

  ```bash
  export HFTOKEN="insert-your-huggingface-token-here"
  ```

### Deployment

#### 1. Deploy with CPUs only

This setup ensures that all services in the opea/chatqna workload run with CPU resources only.

```bash
# Get the list of available charts:
helm search repo opea

# Deploy the opea/chatqna chart
helm install chatqna opea/chatqna --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN}
```
Verify deployment status:
- Deployed Kubernetes services:

  These are the actual services running in the cluster. Use the following command to list their runtime names:
  ```bash
  kubectl get svc
  ```
- Check the status of pods and ensure they are ready before proceeding:
  ```bash
  kubectl get pods
  ```

#### 2. Deploy with Gaudi Devices

To offload TEI embedding and TGI inference to Intel Gaudi devices, use the specific values file.

```bash
# Extract chart contents to access provided configuration files.
helm pull opea/chatqna --untar

# Deploy using the Gaudi-specific configuration.
helm install chatqna opea/chatqna --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} -f chatqna/gaudi-values.yaml 

```

> #### 3. Deploy with Multiple Instances

Reviewer: please see earlier comment about plural ..

Author: s/Deploy with Multiple Instances/Deploy with Multiple Pod Instances

To effectively deploy your workload across multiple nodes, follow these steps.

##### Preload Shared Models

> Donwloading models simultaneously across multiple instances can overload resources. To prevent resource exhaustion, it's recommended to preload the models in advance.

Reviewer: ```suggestion
Downloading models simultaneously to multiple nodes in your cluster can overload network resources. To prevent resource exhaustion, it's recommended to preload the models to a local cache and re-use as illustrated below.
```

Author: Changed as "Downloading models simultaneously to multiple nodes in your cluster can overload resources such as network bandwidth, memory and storage. To prevent resource exhaustion, it's recommended to preload the models in advance."

```bash
pip install -U "huggingface_hub[cli]"
sudo mkdir -p /mnt/models
sudo chmod 777 /mnt/models
huggingface-cli download --cache-dir /mnt/models Intel/neural-chat-7b-v3-3
export MODELDIR=/mnt/models
```
Once the models are downloaded, you can consider the following methods for sharing them across nodes:

> - Persistent Volume Claim (PVC): This is recommended approach for production setups. For more details on using PVC, refer to [PVC](https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md#using-persistent-volume).

Reviewer: ```suggestion
- Persistent Volume Claim (PVC): This is the recommended approach for production setups. For more details on using PVC, refer to [PVC](https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md#using-persistent-volume).
```

Author: The update is donw as suggested.

> - Local Host Path: For simpler testing, ensure that each node involved in the deployment follows the steps above to locally prepare the modes. After preparing the models, use `--set global.modelUseHostPath=${MODELDIR}` in the deployment command.

Reviewer: ```suggestion
- Local Host Path: For simpler testing, ensure that each node involved in the deployment follows the steps above to locally prepare the models. After preparing the models, use `--set global.modelUseHostPath=${MODELDIR}` in the deployment command.
```

```
# cluster.yaml - Example configuration for multiple instances
tgi:                          # Chart-defined service name
  replicaCount: 12            # Number of instances
  evenly_distributed: true    # Distribute instances evenly across hardware nodes
  affinity:                   # Schedule pods on specific nodes
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
          - matchExpressions:
              - key: kubernetes.io/hostname
                operator: In
                values:
                  - node1
                  - node2
```
##### Deploy with Your Custom Configuration
Use the following command to deploy the workload with multiple instances:

```bash
helm install chatqna opea/chatqna \
  --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} \
  --set global.modelUseHostPath=${MODELDIR} \
  -f chatqna/gaudi-values.yaml -f cluster.yaml
```

#### 4. Deploy with a Specific Namespace

To deploy the workload in a specific namespace:

```bash
export NAMESPACE=“insert-your-workload-namespace”
helm install chatqna opea/chatqna \
  --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} \
  -f chatqna/gaudi-values.yaml \

>   --namespace ${NAMESPACE} --create-namespace

Reviewer: is there a problem if the namespace already exists in? Perhaps say "create if it does not exist"

Author: `--create-namespace                           create the release namespace if not present`, Refer to the "helm install --help", there is no need to worry.
